pkg/crosscluster/logical/table_batch_handler.go

```diff
@@ -0,0 +1,239 @@
+	if err != nil {
+		return stats, err
+	}
+
+	stats, err = t.attemptBatch(ctx, refreshedBatch)
+	if err != nil {
+		return tableBatchStats{}, err
+	}
+
+	stats.Add(refreshStats)
+
+	return stats, nil
+}
+
+func (t *tableHandler) attemptBatch(
+	ctx context.Context, batch []decodedEvent,
+) (tableBatchStats, error) {
+	var stats tableBatchStats
+	err := t.db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		for _, event := range batch {
+			switch {
+			case event.isDelete && len(event.prevRow) != 0:
+				stats.deletes++
+				err := t.sqlWriter.DeleteRow(ctx, txn, event.originTimestamp, event.prevRow)
+				if isLwwLoser(err) {
+					stats.kvLwwLosers++
+					continue
+				}
+				if err != nil {
+					return err
+				}
+			case event.isDelete && len(event.prevRow) == 0:
+				stats.tombstoneUpdates++
+				tombstoneUpdateStats, err := t.tombstoneUpdater.updateTombstone(ctx, txn, event.originTimestamp, event.row)
+				if err != nil {
+					return err
+				}
+				stats.kvLwwLosers += tombstoneUpdateStats.kvWriteTooOld
+			case event.prevRow == nil:
+				stats.inserts++
+				err := t.sqlWriter.InsertRow(ctx, txn, event.originTimestamp, event.row)
+				if isLwwLoser(err) {
+					stats.kvLwwLosers++
+					continue
+				}
+				if err != nil {
+					return err
+				}
+			case event.prevRow != nil:
+				stats.updates++
+				err := t.sqlWriter.UpdateRow(ctx, txn, event.originTimestamp, event.prevRow, event.row)
+				if err != nil {
+					return err
+				}
+			default:
+				return errors.AssertionFailedf("unhandled event type: %v", event)
+			}
+		}
+		return nil
+	})
+	if len(batch) == 1 && isLwwLoser(err) {
+		return tableBatchStats{kvLwwLosers: 1}, nil
+	}
+	if err != nil {
+		return tableBatchStats{}, err
+	}
+	return stats, nil
+}
+
+// refreshPrevRows refreshes the prevRow field for each event in the batch. If
+// any event is known to be a lww loser based on the read, its dropped from the
+// batch.
+func (t *tableHandler) refreshPrevRows(
+	ctx context.Context, batch []decodedEvent,
+) ([]decodedEvent, tableBatchStats, error) {
+	var stats tableBatchStats
+	stats.refreshedRows = int64(len(batch))
+
+	rows := make([]tree.Datums, 0, len(batch))
+	for _, event := range batch {
+		rows = append(rows, event.row)
+	}
+
+	var refreshedRows map[int]priorRow
+	err := t.db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		var err error
+		refreshedRows, err = t.sqlReader.ReadRows(ctx, txn, rows)
+		return err
+	})
+	if err != nil {
+		return nil, tableBatchStats{}, err
+	}
+
+	refreshedBatch := make([]decodedEvent, 0, len(batch))
+	for i, event := range batch {
+		refreshed, found := refreshedRows[i]
+
+		revfreshedEvent := decodedEvent{
+			dstDescID:       event.dstDescID,
+			isDelete:        event.isDelete,
+			row:             event.row,
+			originTimestamp: event.originTimestamp,
+		}
+		if found {
+			if !refreshed.logicalTimestamp.Less(event.originTimestamp.Next()) {
+				// Skip the row because it is a lww loser.
```

Comment:
this is an awesome little trick! did not realize we'd get this with the bulk read.

Perhaps we could add a note indicating that we'll still have to replay LWWlosers during delete-delete conflicts.